### PR TITLE
Correct BigNum namespace and Correct Bignum OpenSSL 1.1 Compatibility

### LINKED ISSUE #44: Compile Error because of deprecated BIGNUM in openssl 1.1.0g
BigNum is deprecated.
With openssl 1.1.0 g compiling fails.
Openssl versions prior to 1.1.0 f were working well, but now these errors stop the compiler on Linux

bignum.h:51:24: error: invalid use of incomplete type ‘BIGNUM {aka struct bignum_st}’
 class CBigNum : public BIGNUM
                        ^~~~~~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from allocators.h:12,
                 from crypter.h:7,
                 from keystore.h:8,
                 from script.h:14,
                 from auxpow.cpp:4:
/usr/include/openssl/ossl_typ.h:80:16: note: forward declaration of ‘BIGNUM {aka struct bignum_st}’
 typedef struct bignum_st BIGNUM;
                ^~~~~~~~~
In file included from script.h:15:0,
                 from auxpow.cpp:4:
bignum.h: In constructor ‘CBigNum::CBigNum()’:
bignum.h:56:9: error: ‘BN_init’ was not declared in this scope
         BN_init(this);
         ^~~~~~~
bignum.h:56:9: note: suggested alternative: ‘BN_print’
         BN_init(this);
         ^~~~~~~
         BN_print
bignum.h: In copy constructor ‘CBigNum::CBigNum(const CBigNum&)’:
bignum.h:61:9: error: ‘BN_init’ was not declared in this scope
         BN_init(this);
         ^~~~~~~

### PR DESCRIPTION
**Pull request by ghost**
This problem was encountered by other coins so there was a clue about the fix.
This commit fixes https://github.com/devcoin/core/issues/44

This commit corrects the bignum namespace and openssl 1.1 compatibility issues due to changes in the BIGNUM function in openssl bignum.h

### CODE REVIEW

**Comment by develCuy** [comment]
This fix is based on: https://github.com/peercoin/peercoin/pull/153/commits/5b09830e5de0f5105534e69dbf4acffb3255869b

File: src/bignum.h [hunk 1]
```diff
@@ -560,31 +575,25 @@ inline const CBigNum operator%(const CBigNum& a, const CBigNum& b)
 {
     CAutoBN_CTX pctx;
     CBigNum r;
-    if (!BN_mod(&r, &a, &b, pctx))
+    if (!BN_div(r.get(), NULL, a.cget(), b.cget(), pctx))
```
Comment (develCuy) [inline src/bignum.h]: Copy/paste error. See https://github.com/peercoin/peercoin/pull/153/commits/5b09830e5de0f5105534e69dbf4acffb3255869b line 593

File: src/bignum.h [hunk 2]
```diff
@@ -560,31 +575,25 @@ inline const CBigNum operator%(const CBigNum& a, const CBigNum& b)
...
 inline const CBigNum operator<<(const CBigNum& a, unsigned int shift)
 {
     CBigNum r;
-    if (!BN_lshift(&r, &a, shift))
+    if (!BN_lshift(r.get(), a.cget(), shift))
         throw bignum_error("CBigNum:operator<< : BN_lshift failed");
     return r;
 }
 
-inline const CBigNum operator>>(const CBigNum& a, unsigned int shift)
```
Comment (develCuy) [inline src/bignum.h]: @ctya why do you remove CBugNum? See https://github.com/peercoin/peercoin/pull/153/commits/5b09830e5de0f5105534e69dbf4acffb3255869b line 598

File: src/key.cpp [hunk 3]
```diff
@@ -308,8 +319,15 @@ bool CKey::SignCompact(uint256 hash, std::vector<unsigned char>& vchSig)
...
-    int nBitsR = BN_num_bits(sig->r);
-    int nBitsS = BN_num_bits(sig->s);
+    const BIGNUM *sig_r, *sig_s;
+    #if OPENSSL_VERSION_NUMBER > 0x1000ffffL
+    ECDSA_SIG_get0(sig, &sig_r, &sig_s);
+    #else
+    sig_r = sig->r;
+    sig_s = sig->s;
+    #endif
+     int nBitsR = BN_num_bits(sig_r);
```
Comment (develCuy) [inline src/key.cpp]: Check coding standards. See https://github.com/peercoin/peercoin/pull/153/commits/5b09830e5de0f5105534e69dbf4acffb3255869b line 332

File: src/crypter.cpp [hunk 4]
```diff
@@ -1,4 +1,5 @@
-// Copyright (c) 2009-2012 The Bitcoin Developers
+// Copyright (c) 2009-2012 The Bitcoin developers
```
Comment (develCuy) [inline src/crypter.cpp]: Typo

File: src/crypter.cpp [hunk 5]
```diff
@@ -1,4 +1,5 @@
...
+// Copyright (c) 2009-2018 The Devcoin Developers
```
Comment (develCuy) [inline src/crypter.cpp]: typo

File: src/bignum.h (hunk 2 above)
Comment (ctya) [reply src/bignum.h]: My mistake

**Comment by ctya** [comment]
Thanks for your detailed review!

**Comment by develCuy** [comment]
Great work @ctya!

**Review by DevElCuy — COMMENTED** [review]
Great work!

**Pull request merged**